sklearn/utils/_array_api.py

```diff
@@ -462,28 +499,106 @@ def _weighted_sum(sample_score, sample_weight, normalize=False, xp=None):
             sample_weight_np = None
         return float(numpy.average(sample_score_np, weights=sample_weight_np))
 
-    if not xp.isdtype(sample_score.dtype, "real floating"):
-        # We move to cpu device ahead of time since certain devices may not support
-        # float64, but we want the same precision for all devices and namespaces.
-        sample_score = xp.astype(xp.asarray(sample_score, device="cpu"), xp.float64)
+    if sample_weight is None:
+        sum_ = float(xp.sum(sample_score))
+        n_sample = sample_score.shape[0]
+        if normalize and n_sample != 0:
+            return sum_ / n_sample
+        return sum_
 
-    if sample_weight is not None:
-        sample_weight = xp.asarray(sample_weight, dtype=sample_score.dtype)
-        if not xp.isdtype(sample_weight.dtype, "real floating"):
-            sample_weight = xp.astype(sample_weight, xp.float64)
```

Comment:
> highest precision supported dtype

I'm fine using this, in fact, it's the only possible way forward at the moment, considering there's no mean to convert an array across both namespaces and devices with a standard syntax (see ongoing discussion at https://github.com/data-apis/array-api/issues/710). (another solution could be to generalize sklearn's `_convert_to_numpy` to a wider range of array libraries but that is a fair amount of work).

If we use that for the weighted sum, can we apply the same logic wherever there are explicit casts to float64 as long as it doesn't break tests ? ...what if it does ?

(NB: the discussion about the weighted sum should maybe better take place in https://github.com/scikit-learn/scikit-learn/pull/27232 where @betatim had the initial proposal)

Reply:
I think the question we need to answer is: "should results change or not when switching from one device/array library to another?"

I think generally users assume that if they have a computation in PyTorch that works (doesn't crash, gives sensible results, converges, etc)  on the CPU then the same will be true when they switch to PyTorch on a GPU. This allows you to, for example, develop on your laptop (no GPU) and deploy on a server with a GPU.

I am pretty sure that most people believe that PyTorch makes this kind of promise. I don't know how true it is in practice. I also don't know how easy it is to find out :-/

Before I've said that "If I'm using a device that only supports float32, then of course the results will be different from one that supports float64. It is expected" - but now I think I've changed my mind.

If we can agree on the high level idea "switching array libs/device will not change whether code works or not" (and what "works" means in this context: numerically identical, numerically close given different dtype was used, algorithm continues to converge, algorithm continues to converge to the same output) - then we can use the answer to derive an answer to Olivier's question.

---

On a case by case basis I seem to keep changing my mind about what we should do (use lower precision, convert to numpy, use lower precision if transfering to the CPU is costly, ..) and after thinking a bit more about it I think that answering the higher level question (above) will give us a guide to follow and hence result in consistent decision making.

Reply:
I agree that the choices are best optimized on a case-by-case basis. But it adds work on the PRs and in the short-term, I think it's not worth to do case-by-case, I feel that in many instances it would become a major hurdle, although scikit-learn-like compute with float32-only devices is not something common yet ? and we don't have much data on the negative impacts of float32-only compute ?

So IMO it makes sense to look at a simplistic but consistent policy to begin with, which are either:

- 1) use at most max device precision everywhere
    + pro: no slowdown
    + con: results might be so inaccurate that it's not usable
- 2) or, do not compromise on precision and always copy to cpu when float64 is needed
    + pro: ux and quality is consistent with all devices
    + con: in some cases the overheads might overweight the advantages of doing the rest of the compute on device
- 3) variations: one or both are possible and the user choose with a config keyword, and 1) or 2) is default, or by default and error is raised

and then refine later on ?

Personally I'm ok with all those options, but we should converge on one and be consistent in PRs ? The safe side is clearly 2) but it's not technically feasible unless `_convert_to_numpy` is heavily expanded, or we wait for an uncertain support of this usecase by the Array API

Reply:
> Before I've said that "If I'm using a device that only supports float32, then of course the results will be different from one that supports float64. It is expected" - but now I think I've changed my mind.

If we want to strictly enforce this, we could remove the `float64` upcast from `_weighted_sum`. I am pretty sure it's useless and is a typical case of defensive coding not justified by any numerical stability test in our test suite. But it might risk introducing a behavioral change for some of our users without the Array API enabled. This is why I was suggested to special case numpy in this PR for now, maybe with a TODO to explore removing the `float64` upcast in the future to get consistent behavior across all libraries.

Reply:
My general policy would be:

Try to use the precision of the input data as much as possible and only upcast to `float64` if:

- it is justified by a numerical stability test in our test suite;
- the device allows for it.

Reply:
The alternative would be to use `_convert_to_numpy` to in those cases. 

`_convert_to_numpy` should support all the Array API libraries we officially support in our test suite. We should probably better document that list of libraries in our document: https://scikit-learn.org/dev/modules/array_api.html

Reply:
BTW if we want to move forward quickly and conservatively on this topic, let's just use `_convert_to_numpy` in `_weighted_sum` when we detect that `float64` is not supported by the namespace / device combo.

We can always change do a follow-up PR later to instead stay in float32 land as much as possible for any namespace / device (even the default numpy) but this way we can decouple the discussion of numerical precision in `_weighted_sum` from making progress on adding Array API regression estimators and metrics.

Reply:
> BTW if we want to move forward quickly and conservatively on this topic, let's just use `_convert_to_numpy` in `_weighted_sum` when we detect that `float64` is not supported by the namespace / device combo.

:+1: I'll update the PR in this direction

Reply:
> * the device allows for it.

Yes

> * it is justified by a numerical stability test in our test suite;

this one is more delicate IMO, I'm not convinced that existing tests are enough. Since scikit-learn has assumed float64 is always available so far, contributors might just have conservatively casted to float64 whenever the cost was negligible, without always adding non-regression tests. Some of those cases might be over-cautious, some might be necessary. So at the moment I'd be more in favor of requiring to add more elements to prove that float32 is enough when the usage before that was to enforce float64, beside "it does not break existing tests" ? 

I'd prefer that we can agree to be OK if Array API PRs just conform to the existing requirement regarding precision, even if removing the upcast does not break tests. And leave it to later to optimize if it's possible.

Reply:
> this one is more delicate IMO, I'm not convinced that existing tests are enough.

I meant to add a some new tests if necessary to show that the upcast is necessary. The new tests would server as non-regression numerical stability tests to prove that the upcast to `float64` is justified.

If those tests are not easy to write, it probably means that working in `float32` is good enough and the upcast of to `float64` in the existing codebase was overly defensive.